src/main/java/org/elasticsearch/search/internal/DefaultSearchContext.java

```diff
@@ -197,8 +195,27 @@ public DefaultSearchContext(long id, ShardSearchRequest request, SearchShardTarg
         this.indexShard = indexShard;
         this.indexService = indexService;
 
-        this.searcher = new ContextIndexSearcher(this, engineSearcher);
-
+        aliasFields = indexService.aliasesService().aliasFields(request.filteringAliases());
+        if (aliasFields != null) {
+            // We need to include the meta fields, otherwise features will stop working.
+            Set<String> indexedFieldNames = Sets.newHashSet(MapperService.getMetaFields());
+            // We need to exclude the _all field here, because that includes all a copy of all values from all fields.
+            // which would break the filtering by field
+            indexedFieldNames.remove("_all");
+            Set<String> fullFieldNames = new HashSet<>();
+            for (FieldMapper field : aliasFields) {
+                indexedFieldNames.add(field.names().indexName());
+                fullFieldNames.add(field.names().fullName());
+            }
+            try {
+                DirectoryReader filter = FieldSubsetReader.wrap((DirectoryReader) engineSearcher.searcher().getIndexReader(), indexedFieldNames, fullFieldNames);
+                this.searcher = new ContextIndexSearcher(this, engineSearcher, filter);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```

Comment:
can we just throw ElasticsearchException since it has the HTTP code baked in and it's also a RT exception